Tests/AsyncAlgorithmsTests/TestBuffer.swift

```diff
@@ -193,82 +106,196 @@ final class TestBuffer: XCTestCase {
+//    gated.advance()
+//    gated.advance()
+//    gated.advance()
+//    value = try! await iterator.next()
+//    XCTAssertEqual(value, 2)
+//
+//    gated.advance()
+//    gated.advance()
+//    gated.advance()
+//    gated.advance()
+//    do {
+//      value = try await iterator.next()
+//      XCTFail("next() should have thrown.")
+//    } catch { }
+//
+//    value = try! await iterator.next()
+//    XCTAssertNil(value)
+//
+//    value = try! await iterator.next()
+//    XCTAssertNil(value)
+//
+//    value = try! await iterator.next()
+//    XCTAssertNil(value)
+//
+//    value = try! await iterator.next()
+//    XCTAssertNil(value)
+//
+//    value = try! await iterator.next()
+//    XCTAssertNil(value)
+//  }
+//  
+//  func test_byteBuffer() async {
+//    actor ByteBuffer: AsyncBuffer {
+//      var buffer: [UInt8]?
+//      
+//      func push(_ element: UInt8) async {
+//        if buffer == nil {
+//          buffer = [UInt8]()
+//        }
+//        buffer?.append(element)
+//      }
+//      
+//      func pop() async -> [UInt8]? {
+//        defer { buffer = nil }
+//        return buffer
+//      }
+//    }
+//    
+//    var data = Data()
+//    for _ in 0..<4096 {
+//      data.append(UInt8.random(in: 0..<UInt8.max))
+//    }
+//    let buffered = data.async.buffer {
+//      ByteBuffer()
+//    }
+//    var collected = Data()
+//    for await segment in buffered {
+//      collected.append(contentsOf: segment)
+//    }
+//    XCTAssertEqual(data, collected)
+//  }
+//
   func test_bufferingOldest() async {
     validate {
       "X-12-   34-    5   |"
-      $0.inputs[0].buffer(policy: .bufferingOldest(2))
+      $0.inputs[0].buffer(policy: .bufferingOldest(limit: 2))
       "X,,,[1,],,[2,],[3,][5,]|"
     }
   }
 
   func test_bufferingOldest_noDrops() async {
     validate {
       "X-12   3   4   5   |"
-      $0.inputs[0].buffer(policy: .bufferingOldest(2))
+      $0.inputs[0].buffer(policy: .bufferingOldest(limit: 2))
       "X,,[1,][2,][3,][45]|"
     }
   }
 
   func test_bufferingOldest_error() async {
     validate {
       "X-12345^"
-      $0.inputs[0].buffer(policy: .bufferingOldest(2))
+      $0.inputs[0].buffer(policy: .bufferingOldest(limit: 2))
       "X,,,,,,[12^]"
     }
   }
 
   func test_bufferingNewest() async {
     validate {
       "X-12-   34    -5|"
-      $0.inputs[0].buffer(policy: .bufferingNewest(2))
+      $0.inputs[0].buffer(policy: .bufferingNewest(limit: 2))
       "X,,,[1,],,[3,],[4,][5,]|"
     }
   }
 
   func test_bufferingNewest_noDrops() async {
     validate {
       "X-12   3   4   5   |"
-      $0.inputs[0].buffer(policy: .bufferingNewest(2))
+      $0.inputs[0].buffer(policy: .bufferingNewest(limit: 2))
       "X,,[1,][2,][3,][45]|"
     }
   }
 
   func test_bufferingNewest_error() async {
     validate {
       "X-12345^"
-      $0.inputs[0].buffer(policy: .bufferingNewest(2))
+      $0.inputs[0].buffer(policy: .bufferingNewest(limit: 2))
       "X,,,,,,[45^]"
     }
   }
```

Comment:
It would be good to adapt some cancellation tests

```suggestion
  }
  
  
  func test_buffer_limit_finishes_when_iteration_task_is_cancelled() async {
    let value: String? = "test"
    let source = Indefinite(value: value)
    let sequence = source.async.map {
      try? await Task.sleep(nanoseconds: 1_000_000)
      return $0
    }.buffer(limit: 3)
    let finished = expectation(description: "finished")
    let iterated = expectation(description: "iterated")
    let task = Task {
      var firstIteration = false
      for await _ in sequence {
        if !firstIteration {
          firstIteration = true
          iterated.fulfill()
        }
      }
      finished.fulfill()
    }
    // ensure the other task actually starts
    wait(for: [iterated], timeout: 1.0)
    // cancellation should ensure the loop finishes
    // without regards to the remaining underlying sequence
    task.cancel()
    wait(for: [finished], timeout: 1.0)
  }
  
  func test_buffer_policy_finishes_when_iteration_task_is_cancelled() async {
    let value: String? = "test"
    let source = Indefinite(value: value)
    let sequence = source.async.map {
      try? await Task.sleep(nanoseconds: 1_000_000)
      return $0
    }.buffer(policy: .unbounded)
    let finished = expectation(description: "finished")
    let iterated = expectation(description: "iterated")
    let task = Task {
      var firstIteration = false
      for await _ in sequence {
        if !firstIteration {
          firstIteration = true
          iterated.fulfill()
        }
      }
      finished.fulfill()
    }
    // ensure the other task actually starts
    wait(for: [iterated], timeout: 1.0)
    // cancellation should ensure the loop finishes
    // without regards to the remaining underlying sequence
    task.cancel()
    wait(for: [finished], timeout: 1.0)
  }
```

Reply:
added.